### ISSUE TITLE
Psych breaks on \u0092 character

### ISSUE THREAD

**Issue by JamesChevalier** [ruby/psych#271]
I'm having a hard time with this one & I'm hoping you can help me out...

I've got this `\u0092` character appearing in some yaml that's being written by [audited](https://github.com/collectiveidea/audited). When I attempt to read the yaml that includes this character back, Psych gives the error:

```
Psych::SyntaxError: (<unknown>): control characters are not allowed at line 1 column 1
```

I previously [reported the issue to audited](https://github.com/collectiveidea/audited/issues/260), but now I'm not sure where it should go based on some further research I've done...

I saw that you have two other old issues regarding character encodings: [Psych breaks on \u0084 character](https://github.com/tenderlove/psych/issues/207), and [control characters are not allowed at line 1 column 1](https://github.com/tenderlove/psych/issues/116). I read through those, and saw your `ok? value` method. It doesn't pick up this character, though. I also found [this page](https://rosettacode.org/wiki/Strip_control_codes_and_extended_characters_from_a_string#Ruby) which lists out ways to strip control codes and extended characters from a string in many languages and found that only the `strip_control_and_extended_characters` method worked to remove that character.

I know that Psych is supposed to adhere to the [YAML Character Set Spec](http://www.yaml.org/spec/1.2/spec.html#id2770814), but I don't know enough about character sets to know whether `\u0092` is allowable or not. Based on all of these methods I've tried it seems like it should be allowable, but based on the fact that Psych is crashing it seems like it should not be.

Do you know if this `\u0092` character should be allowed in YAML? Should Psych be able to work with it?

**Comment by JamesChevalier** [comment]
I had one of those half-wake-up-out-of-a-dream-with-an-idea-on-how-to-test-this experiences last night, and it actually **worked**. :open_mouth:

The data that I have been working with is data that was migrated over from MySQL to PostgreSQL. I wondered if it was just legacy data that was experiencing this issue, and I was correct. The only reason that I'm in a situation where this `\u0092` character is because it was migrated over from MySQL. New data that gets pushed through the audited gem does **not** return this error - it's actually storing the character as `\x92`.

I still don't know enough about character sets to know whether or not Psych **should** be able to handle this `\u0092` (or `\x92`) character, but I do know that this isn't an issue for me any longer.

**Comment by tenderlove** [comment]
Hi,

The problem is that `\x92` is not a valid character in Unicode.  YAML files are supposed to be Unicode, so whatever wrote that YAML file did something wrong.

```
irb(main):002:0> "\x92".encoding
=> #<Encoding:UTF-8>
irb(main):003:0> "\x92".valid_encoding?
=> false
irb(main):007:0> puts Psych.dump "\x92".b
--- !binary |-
  kg==
=> nil
irb(main):008:0> x = Psych.load Psych.dump "\x92".b
=> "\x92"
irb(main):009:0>
```

Hope that helps!

**Comment by swordfish444** [comment]
@JamesChevalier Did you ever figure out what was causing this? 

What version of the Audited Gem was causing this? And after what version were you no longer seeing incorrect character sequences in the YAML? 

We are migrating a Rails 3.2 app to Rails 5.1 and I'm seeing similiar errors. Rails 5.1 DelayedJob worker processes are failing to deserialize jobs with control characters that were created in the 3.2 app. Whats even more odd is that these same jobs are understood and processed by the Rails 3.2 workers..


The Rails 3.2 app is using Psych 2.0.3 which is 5 years out-of-date, and the Rails 5.1 app is using v3. @tenderlove could this be the culprit for producing the invalid YAML? Do you know if there were any bug fixes/changes in serialization between 2.0.3 and the latest version?

**Comment by dcecile** [comment]
Here are the values cause libyaml to emit the control character error:

```c
            if (! (value == 0x09 || value == 0x0A || value == 0x0D
                        || (value >= 0x20 && value <= 0x7E)
                        || (value == 0x85) || (value >= 0xA0 && value <= 0xD7FF)
                        || (value >= 0xE000 && value <= 0xFFFD)
                        || (value >= 0x10000 && value <= 0x10FFFF)))
                return yaml_parser_set_reader_error(parser,
                        "control characters are not allowed",
                        parser->offset, value);
```

https://github.com/yaml/libyaml/blob/f8f760f7387d2cc56a2fc7b1be313a3bf3f7f58c/src/reader.c#L411-L415
